### ISSUE TITLE
Composer runs out of memory on latest version

### ISSUE THREAD

**Issue by ElvenSpellmaker** [composer/composer#8981]
Hey there, I did a `composer selfupdate` and it updated fine. However most composer commands now throw an out of memory error at random points.

I think my rollback version is a 2.X version whereas running `composer self-update` seems to drop me down to a 1.X version (is this causing the issue?)

Windows 10
Cygwin

Output of `composer diagnose`:

```
Checking composer.json: OK
Checking platform settings: OK
Checking git settings: OK
Checking http connectivity to packagist: OK
Checking https connectivity to packagist: OK
Checking github.com oauth access: OK
Checking disk free space: OK
Checking pubkeys:
Tags Public Key Fingerprint: 57815BA2 7E54DC31 7ECC7CC5 573090D0  87719BA6 8F3BB723 4E5D42D0 84A14642
Dev Public Key Fingerprint: 4AC45767 E5EC2265 2F0C1167 CBBB8A2B  0C708369 153E328C AD90147D AFE50952
OK
Checking composer version: OK
Composer version: 1.10.7
PHP version: 7.3.7
PHP binary path: /usr/bin/php
OpenSSL version: OpenSSL 1.1.1c  28 May 2019
```

When I run this command:
```
laravel new blog
composer clearcache
composer install
```

I get the following output:
```
/tmp/blog [0] $ composer install
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Package operations: 97 installs, 0 updates, 0 removals
  - Installing doctrine/inflector (2.0.3): Downloading (100%)
  - Installing doctrine/lexer (1.2.1): Downloading (100%)
  - Installing dragonmantank/cron-expression (v2.3.0): Downloading (100%)
  - Installing voku/portable-ascii (1.5.1): Downloading (100%)
  - Installing symfony/polyfill-ctype (v1.17.0): Downloading (100%)
  - Installing phpoption/phpoption (1.7.4): Downloading (100%)
  - Installing vlucas/phpdotenv (v4.1.7): Downloading (connecting...)
Downloading (100%)
  - Installing symfony/css-selector (v5.1.1): Downloading (100%)
  - Installing tijsverkoyen/css-to-inline-styles (2.2.2):Downloading (100%)         )
  - Installing symfony/polyfill-php80 (v1.17.0): Downloading (100%)
  - Installing symfony/polyfill-mbstring (v1.17.0): Downloading (100%)
  - Installing symfony/var-dumper (v5.1.1): Downloading (100%)
  - Installing symfony/deprecation-contracts (v2.1.2): Downloading (100%)
  - Installing symfony/routing (v5.1.1): Downloading (100%)
  - Installing symfony/process (v5.1.1): Downloading (100%)
  - Installing symfony/polyfill-php73 (v1.17.0): Downloading (100%)
  - Installing symfony/polyfill-php72 (v1.17.0): Downloading (100%)
  - Installing symfony/polyfill-intl-idn (v1.17.0): Downloading (100%)
  - Installing symfony/mime (v5.1.1): Downloading (100%)
  - Installing symfony/http-foundation (v5.1.1): Downloading (100%)
  - Installing psr/event-dispatcher (1.0.0): Downloading (100%)
  - Installing symfony/event-dispatcher-contracts (v2.1.2): Downloading (connecting..Downloading (100%)
  - Installing symfony/event-dispatcher (v5.1.1): Downloading (100%)
  - Installing psr/log (1.1.3): Downloading (100%)
  - Installing symfony/error-handler (v5.1.1): Downloading (100%)
  - Installing symfony/http-kernel (v5.1.1): Downloading (100%)
  - Installing symfony/finder (v5.1.1): Downloading (100%)
  - Installing symfony/polyfill-intl-normalizer (v1.17.0): Downloading (connecting...Downloading (100%)
  - Installing symfony/polyfill-intl-grapheme (v1.17.0): Downloading (100%)
  - Installing symfony/string (v5.1.1): Downloading (100%)
  - Installing psr/container (1.0.0): Downloading (100%)
  - Installing symfony/service-contracts (v2.1.2): Downloading (100%)
  - Installing symfony/console (v5.1.1): Downloading (100%)
  - Installing symfony/polyfill-iconv (v1.17.0): Downloading (100%)
  - Installing egulias/email-validator (2.1.17): Downloading (100%)
  - Installing swiftmailer/swiftmailer (v6.2.3): Downloading (100%)
  - Installing ramsey/collection (1.0.1): Downloading (100%)
  - Installing brick/math (0.8.15): Downloading (100%)
  - Installing ramsey/uuid (4.0.1): Downloading (100%)
  - Installing psr/simple-cache (1.0.1): Downloading (100%)
  - Installing opis/closure (3.5.4): Downloading (100%)
  - Installing symfony/translation-contracts (v2.1.2): Downloading (100%)
  - Installing symfony/translation (v5.1.1): Downloading (100%)
  - Installing nesbot/carbon (2.35.0): Downloading (100%)
  - Installing monolog/monolog (2.1.0): Downloading (100%)
  - Installing league/flysystem (1.0.69): Downloading (100%)
  - Installing league/commonmark (1.4.3): Downloading (100%)
  - Installing laravel/framework (v7.15.0): Downloading (100%)
  - Installing fideloper/proxy (4.3.0): Downloading (100%)
  - Installing asm89/stack-cors (v2.0.1): Downloading (100%)
  - Installing fruitcake/laravel-cors (v2.0.1): Downloading (100%)
  - Installing ralouphie/getallheaders (3.0.3): Downloading (100%)
  - Installing psr/http-message (1.0.1): Downloading (100%)
  - Installing guzzlehttp/psr7 (1.6.1): Downloading (100%)
  - Installing guzzlehttp/promises (v1.3.1): Downloading (100%)
  - Installing guzzlehttp/guzzle (6.5.4): Downloading (100%)
  - Installing nikic/php-parser (v4.5.0): Downloading (100%)
  - Installing dnoegel/php-xdg-base-dir (v0.1.1): Downloading (100%)
  - Installing psy/psysh (v0.10.4): Downloading (100%)
  - Installing laravel/tinker (v2.4.0): Downloading (100%)
  - Installing scrivo/highlight.php (v9.18.1.1): Downloading (100%)
  - Installing filp/whoops (2.7.3): Downloading (100%)
  - Installing facade/ignition-contracts (1.0.0): Downloading (100%)
  - Installing facade/flare-client-php (1.3.2): Downloading (100%)
  - Installing facade/ignition (2.0.7): Downloading (100%)
  - Installing fzaninotto/faker (v1.9.1): Downloading (60%)PHP Fatal error:  Out of memory (allocated 14680064) (tried to allocate 2097184 bytes) in phar:///usr/local/bin/composer/src/Composer/Util/RemoteFilesystem.php on line 599

Fatal error: Out of memory (allocated 14680064) (tried to allocate 2097184 bytes) in phar:///usr/local/bin/composer/src/Composer/Util/RemoteFilesystem.php on line 599
```

And I expected this to happen:
```
composer self-update --rollback
composer clearcache
composer install
```

```
/tmp/blog [0] $ composer install
Installing dependencies from lock file (including require-dev)
Verifying lock file contents can be installed on current platform.
Package operations: 32 installs, 0 updates, 0 removals
  - Downloading fzaninotto/faker (v1.9.1)
  - Downloading hamcrest/hamcrest-php (v2.0.0)
  - Downloading mockery/mockery (1.4.0)
  - Downloading nunomaduro/collision (v4.2.0)
  - Downloading webmozart/assert (1.8.0)
  - Downloading phpdocumentor/reflection-common (2.1.0)
  - Downloading phpdocumentor/type-resolver (1.1.0)
  - Downloading phpdocumentor/reflection-docblock (5.1.0)
  - Downloading phpunit/php-token-stream (3.1.1)
  - Downloading sebastian/version (2.0.1)
  - Downloading sebastian/type (1.1.3)
  - Downloading sebastian/resource-operations (2.0.1)
  - Downloading sebastian/recursion-context (3.0.0)
  - Downloading sebastian/object-reflector (1.1.1)
  - Downloading sebastian/object-enumerator (3.0.3)
  - Downloading sebastian/global-state (3.0.0)
  - Downloading sebastian/exporter (3.1.2)
  - Downloading sebastian/environment (4.2.3)
  - Downloading sebastian/diff (3.0.2)
  - Downloading sebastian/comparator (3.0.2)
  - Downloading phpunit/php-timer (2.1.2)
  - Downloading phpunit/php-text-template (1.2.1)
  - Downloading phpunit/php-file-iterator (2.0.2)
  - Downloading theseer/tokenizer (1.1.3)
  - Downloading sebastian/code-unit-reverse-lookup (1.0.1)
  - Downloading phpunit/php-code-coverage (7.0.10)
  - Downloading doctrine/instantiator (1.3.1)
  - Downloading phpspec/prophecy (v1.10.3)
  - Downloading phar-io/version (2.0.1)
  - Downloading phar-io/manifest (1.0.3)
  - Downloading myclabs/deep-copy (1.9.5)
  - Downloading phpunit/phpunit (8.5.5)
  - Installing fzaninotto/faker (v1.9.1): Extracting archive
  - Installing hamcrest/hamcrest-php (v2.0.0): Extracting archive
  - Installing mockery/mockery (1.4.0): Extracting archive
  - Installing nunomaduro/collision (v4.2.0): Extracting archive
  - Installing webmozart/assert (1.8.0): Extracting archive
  - Installing phpdocumentor/reflection-common (2.1.0): Extracting archive
  - Installing phpdocumentor/type-resolver (1.1.0): Extracting archive
  - Installing phpdocumentor/reflection-docblock (5.1.0): Extracting archive
  - Installing phpunit/php-token-stream (3.1.1): Extracting archive
  - Installing sebastian/version (2.0.1): Extracting archive
  - Installing sebastian/type (1.1.3): Extracting archive
  - Installing sebastian/resource-operations (2.0.1): Extracting archive
  - Installing sebastian/recursion-context (3.0.0): Extracting archive
  - Installing sebastian/object-reflector (1.1.1): Extracting archive
  - Installing sebastian/object-enumerator (3.0.3): Extracting archive
  - Installing sebastian/global-state (3.0.0): Extracting archive
  - Installing sebastian/exporter (3.1.2): Extracting archive
  - Installing sebastian/environment (4.2.3): Extracting archive
  - Installing sebastian/diff (3.0.2): Extracting archive
  - Installing sebastian/comparator (3.0.2): Extracting archive
  - Installing phpunit/php-timer (2.1.2): Extracting archive
  - Installing phpunit/php-text-template (1.2.1): Extracting archive
  - Installing phpunit/php-file-iterator (2.0.2): Extracting archive
  - Installing theseer/tokenizer (1.1.3): Extracting archive
  - Installing sebastian/code-unit-reverse-lookup (1.0.1): Extracting archive
  - Installing phpunit/php-code-coverage (7.0.10): Extracting archive
  - Installing doctrine/instantiator (1.3.1): Extracting archive
  - Installing phpspec/prophecy (v1.10.3): Extracting archive
  - Installing phar-io/version (2.0.1): Extracting archive
  - Installing phar-io/manifest (1.0.3): Extracting archive
  - Installing myclabs/deep-copy (1.9.5): Extracting archive
  - Installing phpunit/phpunit (8.5.5): Extracting archive
Generating optimized autoload files
> Illuminate\Foundation\ComposerScripts::postAutoloadDump
> @php artisan package:discover --ansi
PHP Warning:  Module 'readline' already loaded in Unknown on line 0
Discovered Package: facade/ignition
Discovered Package: fideloper/proxy
Discovered Package: fruitcake/laravel-cors
Discovered Package: laravel/tinker
Discovered Package: nesbot/carbon
Discovered Package: nunomaduro/collision
Package manifest generated successfully.
43 packages you are using are looking for funding.
Use the `composer fund` command to find out more!
````

**Comment by ElvenSpellmaker** [comment]
I've raised the mem limit from 128M to 256M, 512M and then 2G with no effects, it dies at the same point with the same error.

I get a similar error when I run `composer create-project laravel/laravel --prefer-dist new-laravel-project` at a different point with a different set of memory limits.

**Comment by Seldaek** [comment]
Out of memory indicates the machine is out of memory to be allocated, so the memory limit won't help, you need swap here. See https://getcomposer.org/doc/articles/troubleshooting.md#proc-open-fork-failed-errors

Using Composer 2.x preview builds via `composer self-update --preview` might help though, as v2 does use a lot less memory.

**Comment by ElvenSpellmaker** [comment]
My laptop has 16GB RAM and I'm not using that much currently. I currently only have a browser and editor running (plus apps in the back) and so I'm definitely not using all that RAM.

Unless composer is using gigabytes of RAM I don't see how I've exhausted the system memory.

I'll check again later to see if I was anywhere near my system limit. =)

**Comment by Seldaek** [comment]
You can run the commands with `--profile` to get more info about memory usage.

**Comment by ElvenSpellmaker** [comment]
Thanks!

**Comment by jrfnl** [comment]
I just ran into the same thing and upgrading to Composer 2.0 isn't an option as the project uses a Composer plugin which isn't compatible (yet) with Composer 2.0.

After some trial and error, largely based on info from when the CI build was still passing, I've found that doing a `composer self-update 1.9.3` fixes it for us.

I haven't analyzed which version of Composer introduced the error between 1.9.3 and the last 1.x release, but hopefully this info may help someone else running into this.

**Comment by ElvenSpellmaker** [comment]
I still get these errors and I'm clearly not running out of memory, there's clearly something wrong here.
![image](https://user-images.githubusercontent.com/2286713/85048457-3e2e1780-b18b-11ea-83e0-c724f76b16f4.png)

**Comment by Seldaek** [comment]
Ah you're on windows. It might be that your PHP is running 32bit then? Check if `composer show -p php-64bit` returns a package or not.

But it doesn't seem to be the problem as the amount of memory allocated is only 40MB there.

What do you run this in tho, is it WSL, WSL2, or some sort of VM?

**Comment by ElvenSpellmaker** [comment]
Hey @Seldaek it's in my original post, I'm using Cygwin. =)

It has full access to all my RAM and it's never been a problem until now.

**Comment by Seldaek** [comment]
Ah missed that. I'd generally speaking highly recommend moving over to WSL which is a lot less hackish than cygwin, but I don't know if it'd fix this.

I am really not sure what is going on here.. But have you checked if it's 64 bit or not? Because IIRC 32bit windows processes are limited to 2GB of memory allocation, and I'm not sure using cygwin and whatnot whether that would include all of cygwin processes or not. Definitely seems like some sort of environment issue and not something we can fix I'm afraid.

**Comment by ElvenSpellmaker** [comment]
I really dislike WSL compared to Cygwin and think WSL is more of a hack haha.

Yeah it's 64-bit.
```
$ composer show -p | grep 64
lib-ICU              64.2     The intl PHP library
php-64bit            7.3.7    The PHP interpreter, 64bit
```

**Comment by ElvenSpellmaker** [comment]
@jrfnl what's your setup like?

**Comment by jrfnl** [comment]
@ElvenSpellmaker For me, the problem only occured in CI with a low PHP version to boot. Full details: https://travis-ci.org/github/Yoast/whip/jobs/699645651

I didn't even attempt to test with that low a PHP version locally ;-)

**Comment by ElvenSpellmaker** [comment]
I can't find a version where a 1.X version now works after having used a 2.X version which is weird.

But it's clearly a non-OS problem if it also affects Linux too.

**Comment by ElvenSpellmaker** [comment]
OK this isn't a composer issue. 🙂 

I can't get over 40MB in any PHP process, I'll try downgrading my PHP version.

**Comment by jrfnl** [comment]
@ElvenSpellmaker Well, that same CI job I mentioned before did pass without problems (and 1.5Gb memory) before, but doesn't now without downgrading Composer, so there is definitely something which has recently changed in Composer 1.x causing memory race conditions in certain circumstances...

**Comment by gmponos** [comment]
hi @jrfnl 

>  did pass without problems before

how long is before.. the reason I am asking is that after some time there might be more versions of packages tagged.. hence composer has a bigger matrix of versions to resolve which consumes more memory/process time.. This an hypothesis I am making and I an not sure.. just throwing it out.. as a possible reason..

**Comment by jrfnl** [comment]
@gmponos I understand where you are coming from , but AFAICS, that's not the problem here.
For that particular build, there is one-non-dev dependency which hasn't had a release since 2016 and one-dev dependency which is at dev-master and there is no conflict between the two.

**Comment by rootpric** [comment]
I am having the same problem. I tried all suggestions on this post but still getting about the same result. I have 2 Gig of mem. it started after I did a kernal update in WHM on Centos Linux Cpanel server. Have no idea what to do now?

Fatal error: Allowed memory size of 1610612736 bytes exhausted (tried to allocate 4096 bytes) in phar:///opt/cpanel/composer/bin/composer/src/Composer/DependencyResolver/RuleSetGenerator.php on line 129

Check https://getcomposer.org/doc/articles/troubleshooting.md#memory-limit-errors for more info on how to handle out of memory errors.[oiwb57pb88ya@ip-132-148-167-96 test]$ composer show -p php-64bit
name     : php-64bit
descrip. : The PHP interpreter, 64bit
keywords :
versions : * 7.4.7
type     : library
source   : []
dist     : []
path     :
names    : php-64bit
